Bottom: b1f20c92fe61f3ed09dd893b46eb0ec181bee9b4
Top:    a287e451b32bda1abf8d58b4111e105c8271b5f4
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-05-12 11:02:11 +0800

vhost_net: introduce vhost_has_more_pkts()

Signed-off-by: Jason Wang <jasowang@redhat.com>

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index de68598..6df0c4a 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -481,8 +481,15 @@ static int init_iov_iter(struct vhost_virtqueue *vq, struct iov_iter *iter,
 
 static bool vhost_exceeds_weight(int pkts, int total_len)
 {
-	return total_len >= VHOST_NET_WEIGHT ||
-	       pkts >= VHOST_NET_PKT_WEIGHT;
+	return unlikely(total_len >= VHOST_NET_WEIGHT) ||
+	       unlikely(pkts >= VHOST_NET_PKT_WEIGHT);
+}
+
+static bool vhost_has_more_pkts(struct vhost_net *net,
+				struct vhost_virtqueue *vq)
+{
+	return !vhost_vq_avail_empty(&net->dev, vq) &&
+	       likely(!vhost_exceeds_maxpend(net));
 }
 
 /* Expects to be always run from workqueue - which acts as
@@ -578,8 +585,7 @@ static void handle_tx(struct vhost_net *net)
 		}
 		total_len += len;
 		if (total_len < VHOST_NET_WEIGHT &&
-		    !vhost_vq_avail_empty(&net->dev, vq) &&
-		    likely(!vhost_exceeds_maxpend(net))) {
+		    vhost_has_more_pkts(net, vq)) {
 			msg.msg_flags |= MSG_MORE;
 		} else {
 			msg.msg_flags &= ~MSG_MORE;
@@ -605,7 +611,7 @@ static void handle_tx(struct vhost_net *net)
 		else
 			vhost_zerocopy_signal_used(net, vq);
 		vhost_net_tx_packet(net);
-		if (unlikely(vhost_exceeds_weight(++sent_pkts, total_len))) {
+		if (vhost_exceeds_weight(++sent_pkts, total_len)) {
 			vhost_poll_queue(&vq->poll);
 			break;
 		}
